File: data_structures/ArrayQueue.rb
class ArrayQueue

	def initialize(size)
		@size = size
		@queue = Array.new(size)
		@front = 0
		@back = 0
		@count = 0
	end

	def enqueue(item)

		if isFull?
			raise 'OVERFLOW!'
		end

		@queue[@back] = item
		@back = (@back + 1) % @size
		@count += 1
	end

	def dequeue

		if isEmpty?
			raise 'UNDERFLOW'
		end

		removed = @queue[@front]
		@queue[@front] = nil

		@front = (@front + 1) % @size
		@count -= 1

		removed
	end

	def peek
		@queue[@front]
	end

	def isEmpty?
		@count == 0
	end

	def isFull?
		@count == @size
	end

	def reverse(q)
		s = []


		while (!q.empty?)
			s.push q.shift
		end

		while (!s.empty?)
			q.push s.pop
		end

		q
	end

	def to_s
		@queue.inspect
	end

end

File: data_structures/ArrayQueue.js
class ArrayQueue {

	constructor(size) {
		this.size = size;
		this.count = 0;
		this.front = 0;
		this.end = 0;
		this.queue = new Array(size);
	}

	isFull() {
		return this.count === this.size;
	}

	isEmpty() {
		return this.count === 0;
	}

	enqueue(item) {

		if (this.isFull())
			throw new Error(`array queue is full, could not insert element ${item}`);
		else {

			this.queue[this.end] = item;
			this.end = (this.end + 1) % this.size;
			this.count ++;
		}
	}

	dequeue() {
		if (this.isEmpty())
			throw new Error("array queue is empty");
		else {

			const dequeued = this.queue[this.front];
			this.queue[this.front] = 0;
			this.front = (this.front + 1) % this.size;
			this.count --;

			return dequeued;
		}
	}

	peek() {
		if (this.isEmpty())
			throw new Error("array queue is empty. Cannot peek");
		else
			return this.queue[this.front];
	}
}

module.exports = ArrayQueue;